### DISCUSSION TITLE
config generation for nvim on windows.

### DISCUSSION THREAD

**Discussion by tincotema** [nix-community/nixvim#1455] [Q&A]
Hi,

so here is my problem:
On all my personal systems i run nixos wiht nixvim as my main (and only) editor.
At work i HAVE to work on Windows 10 due to some enterprice software that only works in windows.
I would realy like to be able to still use Nvim on that windows machine, but i would like to avoid the work of having to maintain two configs.

thus the questions is:
is there a elegant (and preverable easy) way of extracting the compleat config, with plugins from the nixvim config to install it on a windows machine?

**Comment by SergioRibera** [comment]
Nix is not available for windows yet, you could use [NixOS-WSL](https://github.com/nix-community/NixOS-WSL) to have your configuration, also nix can export your config to a docker container using [nixos-generators](https://github.com/nix-community/nixos-generators?tab=readme-ov-file#supported-formats) and maybe it will be more interesting.

I don't know how you have your configuration, but personally what [this guy](https://github.com/nmasur/dotfiles/blob/e43fc0f8db53268c4950d2ee12e661f64194b433) did in his config is genius, he gave me the idea of being able to build each application in the `apps` folder as a docker image and run it wherever you want.

Unfortunately I am only proposing ideas, but this is what I want to do in my config precisely because of a similar problem.

**Comment by GaetanLepage** [comment]
I don't have a lot more to add on top of @SergioRibera's answer.
If installing and using WSL is feasible on your windows system, then I think that it is the best option. (You can use [NixOS-WSL](https://github.com/nix-community/NixOS-WSL) for that).
If you have to use the windows version of neovim, then most of what nixvim does (plugin and dependency installation) makes no sense (except the generation of the `init.lua` of course).

**Reply by SergioRibera** [reply]
Right, I forgot to add the link to NixOS-WSL, thank you very much

**Comment by Nibodhika** [comment]
Hi, I started using nixvim recently and have a similar situation. In case anyone finds this post in the future here's how I went about it. Foreword: This is NOT ideal, and will be a pain in the ass to maintain, BUT it allows you to keep nixvim as your main thing on your main machine and provides a way to make it work on Windows.

# Step 1: Choose a package manager
Since windows doesn't have nix available we need to use a different package manager to get our plugins, I went with lazy, just because it was what I was using until switching to nixvim. But similar approach should work with any other package manager.

# Step 2: Setup lazy.nvim
I followed the steps here: https://lazy.folke.io/installation which essentially ends up with a very basic `init.lua` that only includes a file a file for lazy which in turn includes all files in a plugin folder. Now we can write our main plugin definition. For the time being create a plugin (I called mine `all.lua`) and have it be:

```lua
return {
}
```

You should have a working nvim now.

# Step 3: Copy init.lua from your machine
Copy the init.lua nixvim generated in your machine to some place inside the config folder, e.g. `nvim/lua/config/nixvim.lua` and include it in your init.lua file after lazy. So your init.lua file should look something like like:
```lua
require("config.lazy")
require("config.nixvim")
```

# Step 4: Install plugins
This is the boring part, open nvim, it will complain about a plugin not being installed, add it to the `plugins/all.lua` file, don't need to configure it, that will be done by the `nixvim.lua` file, this just ensures it's installed. Rince and repeat until the config stops complaining, e.g. this is a sample of my `all.lua` file:
```lua
return {
	{ "nvim-tree/nvim-web-devicons" },
	{ "nvim-mini/mini.icons" },
	{ "nvim-mini/mini.base16" },
...
        -- some stuff has dependencies:
	{ "kevinhwang91/nvim-ufo", dependencies = 'kevinhwang91/promise-async' },
	{ "nvim-neotest/neotest",
		dependencies = {
			"nvim-neotest/nvim-nio",
			"nvim-lua/plenary.nvim",
			"antoinemadec/FixCursorHold.nvim",
			"nvim-treesitter/nvim-treesitter"
		}
	},
...
}
```

# Step 5: LSG
Haven't gotten to this yet, but LSGs will be a PitA to solve, probably the best idea here is to have a dedicated file for Mason so you can install the required stuff, and again leave the configuration to nixvim.

---

Hope this helps someone.

**Comment by alexandru0-dev** [comment]
i actually would love this feature as also at work they are forcing me to go from nixos to windows
@GaetanLepage i've seen the work with the lazy.nvim plugin manager, will it support installing packages from lazy.nvim?
From the options i can see a dependencies attr, idk if that's related tho.
because if so could be used to generate configs os-indipendent
 yes, we still would lack mason or something like that, but it's a start

**Reply by GaetanLepage** [reply]
Yes, `nixvim` allows you to configure the `lazy.nvim` plugin manager (see [docs](https://nix-community.github.io/nixvim/plugins/lazy/index.html)).

Have a look at [this test](https://nix-community.github.io/nixvim/plugins/lazy/index.html) for how to use it in Nix.
If you want to use `lazy.nvim`'s traditional plugin downloading mechanism, you can directly use `plugins.lazy`'s `settings` option.

**Reply by alexandru0-dev** [reply]
I've tried using the `plugins` attr but it seems to me that this only works with nix packages.
It's ok if the cmd, names and dependencies etc, are populated by the nix package by default if present, but if not present in nixpkgs there is no way to add arbitrary entries since `pkg` is a required attr, also dir is always populated with the nix path even when using `lib.mkForce null` on `dir` (making it so the nix path is needed, which defeats the purpose of having config generation)

let me know if i got something wrong

BTW. the link of tests is wrong but it's ok, i have no issues going around the repo and tests etc

**Reply by GaetanLepage** [reply]
Sorry, here is the link I wanted to send: https://github.com/nix-community/nixvim/blob/5da380555961da4f06fd88cccea218db3558def4/tests/test-sources/plugins/pluginmanagers/lazy.nix#L8-L91.

Indeed, when you use the `plugins`, you work with nix packages.
You can directly set-up `lazy.nvim` using the `plugins.lazy.settings` option. This is directly passed to `require("lazy").setup()` without any alteration.

**Reply by alexandru0-dev** [reply]
> Sorry, here is the link I wanted to send:

Thxx (already found it, but good for other to keep track)

> Indeed, when you use the `plugins`, you work with nix packages. You can directly set-up `lazy.nvim` using the `plugins.lazy.settings` option. This is directly passed to `require("lazy").setup()` without any alteration.

Would it make sense to make `pkg` optional here? (shared interface for nixvim and lazyvim installation)
or a way to override `dir`? (so it can be forced to be null of non-unix targets like config generation)

> You can directly set-up lazy.nvim using the plugins.lazy.settings option. This is directly passed to require("lazy").setup() without any alteration.

i will give it a try

**Reply by GaetanLepage** [reply]
> Would it make sense to make pkg optional here? (shared interface for nixvim and lazyvim installation)
or a way to override dir? (so it can be forced to be null of non-unix targets like config generation)

If you use directly `plugins.lazy.settings`, Nixvim will not prefill anything. You can provide whatever you want. The rest will be ignored of course.

> i will give it a try

How did it go?

**Comment by mairs8** [comment]
see https://github.com/nix-community/nixvim/issues/3738 for an alternate approach. It uses `nixvim.content` and new `vim.pack.add` api.
